File: apps/mobile/lib/features/assistant/assistant_perf.dart
/// 小爱单次 SSE 流式关键路径耗时（客户端 RUM，对齐 Web assistant_perf.ts）。
library;

import 'assistant_visible.dart';

class AssistantPerfDetail {
  const AssistantPerfDetail({this.surface, this.scene});

  final String? surface;
  final String? scene;
}

class AssistantStreamPerf {
  AssistantStreamPerf([this.detail = const AssistantPerfDetail()]);

  final AssistantPerfDetail detail;
  final int _t0 = DateTime.now().millisecondsSinceEpoch;

  int? _tPlaceholderMeta;
  int? _tFullMeta;
  int? _tFirstToken;
  int? _tVisible;
  bool _finished = false;

  static const _maxMarks = 40;
  static final List<_PerfMark> _buffer = [];

  void onPlaceholderMeta() {
    _tPlaceholderMeta ??= _now();
  }

  void onFullMeta({
    bool? cacheHit,
    String? cacheSource,
    int? prepareMs,
  }) {
    _tFullMeta ??= _now();
    if (prepareMs != null && prepareMs >= 0) {
      _record('assistant.prepare', prepareMs, {
        ..._detailMap,
        'cacheHit': cacheHit,
        'cacheSource': cacheSource,
        'source': 'server',
      });
    } else if (_tPlaceholderMeta != null && _tFullMeta != null) {
      _record('assistant.prepare', _tFullMeta! - _tPlaceholderMeta!, {
        ..._detailMap,
        'cacheHit': cacheHit,
        'cacheSource': cacheSource,
        'source': 'client',
      });
    }
    if (cacheHit == true) {
      _record('assistant.cache_hit', _now() - _t0, {
        ..._detailMap,
        'cacheSource': cacheSource,
      });
    }
  }

  void onFirstToken() {
    if (_tFirstToken != null) return;
    _tFirstToken = _now();
    _record('assistant.first_token', _tFirstToken! - _t0, _detailMap);
  }

  void onTextUpdate(String text) {
    if (_tVisible != null) return;
    if (!hasVisibleAnswerContent(text)) return;
    _tVisible = _now();
    _record('assistant.visible', _tVisible! - _t0, _detailMap);
  }

  void onDone() {
    if (_finished) return;
    _finished = true;
    final doneAt = _now();
    _record('assistant.done', doneAt - _t0, _detailMap);
    if (_tFirstToken != null) {
      _record('assistant.stream_body', doneAt - _tFirstToken!, _detailMap);
    }
  }

  void onError() {
    _finished = true;
    _record('assistant.error', _now() - _t0, _detailMap);
  }

  Map<String, Object?> get _detailMap => {
        if (detail.surface != null) 'surface': detail.surface,
        if (detail.scene != null) 'scene': detail.scene,
      };

  int _now() => DateTime.now().millisecondsSinceEpoch;

  static void _record(
    String name,
    int ms, [
    Map<String, Object?>? detail,
  ]) {
    if (ms < 0) return;
    _buffer.add(_PerfMark(name, ms, DateTime.now().millisecondsSinceEpoch, detail));
    if (_buffer.length > _maxMarks) {
      _buffer.removeAt(0);
    }
    assert(() {
      // ignore: avoid_print
      print('[perf] $name ${ms}ms ${detail ?? ''}');
      return true;
    }());
  }

  /// 调试/测试读取最近打点。
  static List<Map<String, Object?>> recentMarks() => [
        for (final m in _buffer)
          {
            'name': m.name,
            'ms': m.ms,
            'at': m.at,
            if (m.detail != null) ...m.detail!,
          },
      ];
}

class _PerfMark {
  _PerfMark(this.name, this.ms, this.at, this.detail);
  final String name;
  final int ms;
  final int at;
  final Map<String, Object?>? detail;
}

File: apps/web/lib/assistant_perf.ts
import { hasVisibleAnswerContent } from '@/lib/assistant_visible';
import { API_BASE, authHeaders } from '@/lib/api';
import { getDeviceId } from '@/lib/device_id';
import { recordPerf } from '@/lib/perf_rum';

export type AssistantPerfDetail = {
  surface?: string;
  scene?: string;
  cacheHit?: boolean;
  cacheSource?: string;
};

type SessionMark = {
  name: string;
  ms: number;
  detail?: Record<string, string | number | boolean | null | undefined>;
};

/** 小爱单次 SSE 流式关键路径耗时（客户端 RUM）。 */
export class AssistantStreamPerf {
  private readonly t0 =
    typeof performance !== 'undefined' ? performance.now() : Date.now();

  private tPlaceholderMeta?: number;
  private tFullMeta?: number;
  private tFirstToken?: number;
  private tVisible?: number;
  private finished = false;
  private readonly sessionMarks: SessionMark[] = [];

  constructor(private readonly detail: AssistantPerfDetail = {}) {}

  onPlaceholderMeta(): void {
    if (this.tPlaceholderMeta == null) {
      this.tPlaceholderMeta = this.now();
    }
  }

  onFullMeta(meta?: {
    cache_hit?: boolean;
    cache_source?: string;
    timings?: { prepare_ms?: number };
  }): void {
    if (this.tFullMeta == null) {
      this.tFullMeta = this.now();
    }
    const serverPrepare = meta?.timings?.prepare_ms;
    if (typeof serverPrepare === 'number' && serverPrepare >= 0) {
      this.recordSession('assistant.prepare', serverPrepare, {
        cacheHit: meta?.cache_hit,
        cacheSource: meta?.cache_source,
        source: 'server',
      });
    } else if (this.tPlaceholderMeta != null && this.tFullMeta != null) {
      this.recordSession('assistant.prepare', this.tFullMeta - this.tPlaceholderMeta, {
        cacheHit: meta?.cache_hit,
        cacheSource: meta?.cache_source,
        source: 'client',
      });
    }
    if (meta?.cache_hit) {
      this.recordSession('assistant.cache_hit', this.now() - this.t0, {
        cacheSource: meta.cache_source,
      });
    }
  }

  onFirstToken(): void {
    if (this.tFirstToken != null) return;
    this.tFirstToken = this.now();
    this.recordSession('assistant.first_token', this.tFirstToken - this.t0);
  }

  onTextUpdate(text: string): void {
    if (this.tVisible != null) return;
    if (!hasVisibleAnswerContent(text)) return;
    this.tVisible = this.now();
    this.recordSession('assistant.visible', this.tVisible - this.t0);
  }

  onDone(): void {
    if (this.finished) return;
    this.finished = true;
    const doneAt = this.now();
    this.recordSession('assistant.done', doneAt - this.t0);
    if (this.tFirstToken != null) {
      this.recordSession('assistant.stream_body', doneAt - this.tFirstToken);
    }
    void flushAssistantPerf(this.sessionMarks);
  }

  onError(): void {
    if (this.finished) return;
    this.finished = true;
    this.recordSession('assistant.error', this.now() - this.t0);
    void flushAssistantPerf(this.sessionMarks);
  }

  private recordSession(
    name: string,
    ms: number,
    extra?: SessionMark['detail'],
  ): void {
    recordPerf(name, ms, { ...this.detail, ...extra });
    if (!Number.isFinite(ms) || ms < 0) return;
    this.sessionMarks.push({
      name,
      ms: Math.round(ms),
      detail: { ...this.detail, ...extra },
    });
  }

  private now(): number {
    return typeof performance !== 'undefined' ? performance.now() : Date.now();
  }
}

/** 将本轮小爱 RUM 打点 batch 上报服务端（fail-open）。 */
export async function flushAssistantPerf(marks: SessionMark[]): Promise<void> {
  if (typeof window === 'undefined' || !marks.length) return;
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...authHeaders(),
  };
  const deviceId = getDeviceId();
  if (deviceId) headers['X-Guest-Id'] = deviceId;
  try {
    await fetch(`${API_BASE}/ai/perf`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ marks: marks.slice(-12) }),
      keepalive: true,
    });
  } catch {
    /* fail-open */
  }
}
